File: dotnet/src/Microsoft.AutoGen/AgentChat/GroupChat/Events.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Events.cs

using Microsoft.AutoGen.AgentChat.Abstractions;

// using ProtobufTypeMarshal = Microsoft.AutoGen.AgentChat.WireProtocol.ProtobufTypeMarshal;

namespace Microsoft.AutoGen.AgentChat.GroupChat;

public class GroupChatEventBase /*: IWireable*/
{
    // public IMessage ToWire()
    // {
    //     return this switch
    //     {
    //         GroupChatStart groupChatStart => ProtobufTypeMarshal.Convert<GroupChatStart, WireProtocol.GroupChatStart>(groupChatStart),
    //         GroupChatAgentResponse groupChatAgentResponse => ProtobufTypeMarshal.Convert<GroupChatAgentResponse, WireProtocol.GroupChatResponse>(groupChatAgentResponse),
    //         GroupChatRequestPublish groupChatRequestPublish => ProtobufTypeMarshal.Convert<GroupChatRequestPublish, WireProtocol.GroupChatRequestPublish>(groupChatRequestPublish),
    //         GroupChatMessage groupChatMessage => ProtobufTypeMarshal.Convert<GroupChatMessage, WireProtocol.GroupChatMessage>(groupChatMessage),
    //         GroupChatTermination groupChatTermination => ProtobufTypeMarshal.Convert<GroupChatTermination, WireProtocol.GroupChatTermination>(groupChatTermination),
    //         GroupChatReset groupChatReset => ProtobufTypeMarshal.Convert<GroupChatReset, WireProtocol.GroupChatReset>(groupChatReset),
    //         _ => throw new InvalidOperationException($"Unknown type {this.GetType().Name}"),
    //     };
    // }
}

/// <summary>
/// A request to start a group chat.
/// </summary>
public class GroupChatStart : GroupChatEventBase
{
    /// <summary>
    /// An optional list of messages to start the group chat.
    /// </summary>
    public List<ChatMessage>? Messages { get; set; }
}

/// <summary>
/// A response published to a group chat.
/// </summary>
public class GroupChatAgentResponse : GroupChatEventBase
{
    /// <summary>
    /// The response from a agent.
    /// </summary>
    public required Response AgentResponse { get; set; }
}

/// <summary>
/// A request to publish a message to a group chat.
/// </summary>
public class GroupChatRequestPublish : GroupChatEventBase
{
}

/// <summary>
/// A message from a group chat.
/// </summary>
public class GroupChatMessage : GroupChatEventBase
{
    /// <summary>
    /// The message that was published.
    /// </summary>
    public required AgentMessage Message { get; set; }
}

/// <summary>
/// A message indicating that group chat was terminated.
/// </summary>
public class GroupChatTermination : GroupChatEventBase
{
    /// <summary>
    /// The stop message that indicates the reason of termination.
    /// </summary>
    public required StopMessage Message { get; set; }
}

/// <summary>
/// A request to reset the agents in the group chat.
/// </summary>
public class GroupChatReset : GroupChatEventBase
{
}

File: typescript/src/agentchat/abstractions/Events.ts
import { AgentMessage, ChatMessage, StopMessage } from "../messages/Messages";
import { Response } from "./ChatAgent";

/**
 * Base class for all group chat events.
 */
export abstract class GroupChatEventBase {}

/**
 * A request to start a group chat.
 */
export class GroupChatStart extends GroupChatEventBase {
    /**
     * An optional list of messages to start the group chat.
     */
    messages?: ChatMessage[];
}

/**
 * A response published to a group chat.
 */
export class GroupChatAgentResponse extends GroupChatEventBase {
    /**
     * The response from an agent.
     */
    agentResponse!: Response;
}

/**
 * A request to publish a message to a group chat.
 */
export class GroupChatRequestPublish extends GroupChatEventBase {}

/**
 * A message from a group chat.
 */
export class GroupChatMessage extends GroupChatEventBase {
    /**
     * The message that was published.
     */
    message!: AgentMessage;
}

/**
 * A message indicating that group chat was terminated.
 */
export class GroupChatTermination extends GroupChatEventBase {
    /**
     * The stop message that indicates the reason of termination.
     */
    message!: StopMessage;
}

/**
 * A request to reset the agents in the group chat.
 */
export class GroupChatReset extends GroupChatEventBase {}
